flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractKeyedStateBackend.java

```diff
@@ -94,52 +89,55 @@
 	/** Decorates the input and output streams to write key-groups compressed. */
 	protected final StreamCompressionDecorator keyGroupCompressionDecorator;
 
+	/** The key context for this backend. */
+	protected final InternalKeyContext<K> keyContext;
+
 	public AbstractKeyedStateBackend(
 		TaskKvStateRegistry kvStateRegistry,
 		TypeSerializer<K> keySerializer,
 		ClassLoader userCodeClassLoader,
-		int numberOfKeyGroups,
-		KeyGroupRange keyGroupRange,
 		ExecutionConfig executionConfig,
 		TtlTimeProvider ttlTimeProvider,
-		CloseableRegistry cancelStreamRegistry) {
+		CloseableRegistry cancelStreamRegistry,
+		InternalKeyContext<K> keyContext) {
 		this(
 			kvStateRegistry,
 			StateSerializerProvider.fromNewRegisteredSerializer(keySerializer),
 			userCodeClassLoader,
-			numberOfKeyGroups,
-			keyGroupRange,
 			executionConfig,
 			ttlTimeProvider,
 			cancelStreamRegistry,
-			determineStreamCompression(executionConfig)
+			determineStreamCompression(executionConfig),
+			keyContext
 		);
 	}
 
 	public AbstractKeyedStateBackend(
 		TaskKvStateRegistry kvStateRegistry,
 		StateSerializerProvider<K> keySerializerProvider,
 		ClassLoader userCodeClassLoader,
-		int numberOfKeyGroups,
-		KeyGroupRange keyGroupRange,
 		ExecutionConfig executionConfig,
 		TtlTimeProvider ttlTimeProvider,
 		CloseableRegistry cancelStreamRegistry,
-		StreamCompressionDecorator keyGroupCompressionDecorator) {
+		StreamCompressionDecorator keyGroupCompressionDecorator,
+		InternalKeyContext<K> keyContext) {
+		Preconditions.checkNotNull(keyContext);
+		this.numberOfKeyGroups = keyContext.getNumberOfKeyGroups();
+		this.keyGroupRange = keyContext.getKeyGroupRange();
+		Preconditions.checkNotNull(keyGroupRange);
```

Comment:
nit: this can simply be `this.keyGroupRange = Preconditions.checkNotNull(keyContext.getKeyGroupRange())`